Tests/LeafKitTests/LeafTests.swift

```diff
@@ -365,6 +365,59 @@ final class LeafTests: XCTestCase {
 //            ])
 //        try XCTAssertEqual(renderer.testRender(template, data), expected)
 //    }
+    
+    func testLoopIndices() throws {
+        let template = """
+        #for(name in names):
+            #(name) - index=#(index) last=#(isLast) first=#(isFirst)
+        #endfor
+        """
+        let expected = """
+
+            tanner - index=0 last=false first=true
+
+            ziz - index=1 last=false first=false
+
+            vapor - index=2 last=true first=false
+
+        """
+    
+        try XCTAssertEqual(render(template, ["names": ["tanner", "ziz", "vapor"]]), expected)
+    }
+    
+    func testNestedLoopIndices() throws {
+        let template = """
+        #for(array in arrays):
+        Array#(index) - [#for(element in array): #(index)#if(isFirst):(first)#elseif(isLast):(last)#endif : "#(element)"#if(!isLast):, #endif#endfor]#endfor
+        """
+        let expected = """
+
+        Array0 - [ 0(first) : "zero",  1 : "one",  2(last) : "two"]
+        Array1 - [ 0(first) : "a",  1 : "b",  2(last) : "c"]
+        Array2 - [ 0(first) : "red fish",  1 : "blue fish",  2(last) : "green fish"]
+        """
+        
+        let data = LeafData.array([
+            LeafData.array(["zero", "one", "two"]),
+            LeafData.array(["a", "b", "c"]),
+            LeafData.array(["red fish", "blue fish", "green fish"])
+        ])
+        
+        try XCTAssertEqual(render(template, ["arrays": data]), expected)
+    }
+    
+    // It would be nice if a pre-render phase could catch things like calling
+    // tags that would normally ALWAYS throw in serializing (eg, calling index
+    // when not in a loop) so that warnings can be provided and AST can be minimized.
+    func testLoopTagsInvalid() throws {
+        let template = """
+            #if(isFirst):Wrong#else:Right#endif
```

Comment:
Is this now invalid if an `isFirst` variable appears in the input context? If so, it should be documented; if not, you should add another test proving that case still works.

Reply:
Good question. I need to double check whether serializer will prefer tags to context... see my comment here for related thoughts https://github.com/vapor/leaf-kit/issues/35#issuecomment-613658779

Reply:
... and it does currently prefer context keys to tag keys when serializing, so I think this is also worth evaluating how things scope.

Reply:
**NB** Above comments should not be taken as **expected** behavior from Leaf and almost certainly is NOT how things should work

Reply:
Ignore all of the above as it seems to have been a dirty build that led me to think the tags were needed; though point about overloading still stands - the loop index of that name will still overload context keys.